### PR TITLE
Handle list random access patterns in frontend

### PR DESCRIPTION
**Pull request by dwightguth**
This PR adds support for the `L [ I <- X ]` syntax on the LHS of rules, which allows you to pattern match on an arbitrary element of a list.

The code is not fully polished; it is not integrated with configuration abstraction or cell collections. It is not likely to be further improved upon presently because the code it was intended to support ended up not proving worth merging. However, it is broadly useful functionality (along with the accompanying LLVM backend PR) and thus ought to be merged.

### CODE REVIEW

**Comment by ehildenb** [comment]
Please add tests to `regression-new` of the new functionality.

File: k-distribution/tests/regression-new/list-set/test.k [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  imports LIST
+  imports INT
+ 
+  configuration <k> $PGM:K </k> <list> ListItem(0) ListItem(1) ListItem(2) </list>
+
+  syntax KItem ::= l(Int, Int)
+
+  rule <k> l(I, J) => .K ...</k>
+       <list> _ [ I <- J ] </list>
+endmodule
```
Comment (ehildenb) [inline k-distribution/tests/regression-new/list-set/test.k:11]: Can we add a test `l(5, 2)` or the like? Something out of range?

**Comment by ehildenb** [comment]
@Scott-Guest will have to comment on the frontend code, but the tests look good, and the feature makes sense.

**Review by Scott-Guest — APPROVED** [review]
Code LGTM, but please
- Document this in "Collection patterns" part of the user manual
- Submit an issue describing the remaining work needed for configuration abstraction and cell collections

**Comment by dwightguth** [comment]
https://github.com/runtimeverification/k/issues/4640
